File: app/src/main/java/com/kinvey/sample/statusshare/model/UpdateEntity.java
/** 
 * Copyright (c) 2014 Kinvey Inc.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 * 
 */
package com.kinvey.sample.statusshare.model;

import java.text.ParsePosition;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Log;

import com.google.api.client.util.Key;
import com.kinvey.android.Client;
import com.kinvey.java.linkedResources.LinkedGenericJson;
import com.kinvey.java.model.KinveyMetaData;
import com.kinvey.java.model.KinveyReference;
import com.kinvey.sample.statusshare.utils.Constants;
import com.kinvey.sample.statusshare.ui.MainActivity;

/**
 * This class maintains a Status Update which can be persisted with Kinvey.
 *
 * @author edwardf
 * @since 2.0
 */
public class UpdateEntity extends LinkedGenericJson {

    //----persisted fields
    @Key("_id")
    private String id;
    @Key("text")
    private String text;
    @Key(Constants.JSON_FIELD_NAME)
    private KinveyMetaData meta;
    @Key("_acl")
    private KinveyMetaData.AccessControlList acl;
    @Key("author")
    private KinveyReference author;
    @Key("comments")
    private ArrayList<KinveyReference> comments;

    public static final int MAX_W = 512;
    public static final int MAX_H = 512;
    public static final String attachmentName = "attachment";

    //-----displayed inferred fields
    private String authorName;
    private String authorID;
    private String since;
    private Bitmap thumbnail;

    public UpdateEntity(){
        putFile(attachmentName);
    }


    public UpdateEntity(String userid) {
        id = null;
        text = null;
        meta = new KinveyMetaData();
        acl = new KinveyMetaData.AccessControlList();
        putFile(attachmentName);
        author = new KinveyReference();
        author.setCollection(Constants.USER_COLLECTION_NAME);
        author.setId(userid);
    }

    public void setAuthor(KinveyReference author){
        this.author = author;
    }

    public KinveyReference getAuthor(){
        return this.author;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getText() {
        return text;
    }

    public void setText(String text) {
        this.text = text;
    }

    public KinveyMetaData.AccessControlList getAcl() {
        return acl;
    }

    public void setAcl(KinveyMetaData.AccessControlList acl) {
        this.acl = acl;
    }

    public KinveyMetaData getMeta() {
        return meta;
    }

    public void setMeta(KinveyMetaData meta) {
        this.meta = meta;
    }

    public String getAuthorID(){
        if (this.author != null && this.author.getResolvedObject() != null){
            this.authorID = (String) (this.author.getResolvedObject()).get("_id");
        }
        return ((this.authorID == null) ? "" : this.authorID);
    }

    public String getAuthorName() {
        if (this.author != null && this.author.getResolvedObject() != null){
            this.authorName = (String) (this.author.getResolvedObject()).get("username");
        }
        return ((this.authorName == null) ? "--" : this.authorName);
    }

    public String getSince() {
        ParsePosition pp = new ParsePosition(0);
        Date date = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS", Locale.US).parse(this.meta.getEntityCreationTime(), pp);
        since = MainActivity.getSince(date, Calendar.getInstance());

        Log.i(Client.TAG,  "getting since -> " + (since != null));
        return since;
    }

    /**
     * Get the thumbnail from the LinkedResource
     *
     * Note it closes the output stream.
     *
     * @return null or the image attachment
     */
    public Bitmap getThumbnail() {
        //If it hasn't been resolved...
        if (thumbnail == null) {
            //and there is an actual LinkedFile behind the Key
            if (getFile(attachmentName) != null) {
                //Then decode from the output stream and get the image.
                thumbnail = BitmapFactory.decodeByteArray(getFile(attachmentName).getOutput().toByteArray(),
                        0, getFile(attachmentName).getOutput().toByteArray().length);
                try {
                    //close the output stream
                    getFile(attachmentName).getOutput().close();
                } catch (Exception e) {
                }
            }
        }
        return thumbnail;
    }

    public ArrayList<KinveyReference> getComments() {
        return comments;
    }

    public void setComments(ArrayList<KinveyReference> comments) {
        this.comments = comments;
    }

    public void addComment(CommentEntity newComment){
        if (this.comments == null){
            this.comments = new ArrayList<KinveyReference>();
        }
        KinveyReference ret = new KinveyReference(Constants.COL_COMMENTS, newComment.getId());
        this.comments.add(ret);
    }

    public void resetCommentReferences(){
        if (this.comments == null){
             return;
        }
        for (KinveyReference c : comments){
            c.remove("_obj");

        }
    }
}

File: app/src/main/java/com/kinvey/sample/statusshare/model/UpdateEntity.kt
/**
 * Copyright (c) 2019 Kinvey Inc.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 *
 */
package com.kinvey.sample.statusshare.model

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import com.google.api.client.util.Key
import com.kinvey.java.linkedResources.LinkedGenericJson
import com.kinvey.java.model.KinveyMetaData
import com.kinvey.java.model.KinveyMetaData.AccessControlList
import com.kinvey.java.model.KinveyReference
import com.kinvey.sample.statusshare.utils.Constants
import com.kinvey.sample.statusshare.utils.Constants.KMD_FIELD_NAME
import com.kinvey.sample.statusshare.utils.Constants.USERNAME_FIELD_NAME
import com.kinvey.sample.statusshare.utils.TimeUtil.getSince
import timber.log.Timber
import java.text.SimpleDateFormat
import java.util.*

/**
 * This class maintains a Status Update which can be persisted with Kinvey.
 *
 * @author edwardf
 * @since 2.0
 */
data class UpdateEntity(
    //----persisted fields
    @Key("_id")
    var id: String? = null,
    @Key("text")
    var text: String? = null,
    @Key(Constants.JSON_FIELD_NAME)
    var meta: KinveyMetaData? = KinveyMetaData(),
    @Key("_acl")
    var acl: AccessControlList? = AccessControlList(),
    @Key("author")
    var author: KinveyReference? = null,
    @Key("comments")
    var comments: ArrayList<KinveyReference>? = null,
    var userId: String? = ""
) : LinkedGenericJson() {

    private var dateFormat: SimpleDateFormat? = null

    init {
        dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS", Locale.US)
        putFile(Constants.ATTACHMENT_NAME)
        author = KinveyReference()
        author?.collection = Constants.USER_COLLECTION_NAME
        author?.id = userId
    }

    //-----displayed inferred fields
    var authorName: String? = null
        get() {
            if (author != null) {
                field = author!![USERNAME_FIELD_NAME] as String?
            }
            return if (field == null) "--" else field
        }
        set (value) {
            if (author != null) {
                author!![USERNAME_FIELD_NAME] = value
            }
        }

    var authorID: String? = null
        get() {
            field = author?.id
            return if (field == null) "" else field
        }
        private set

    var since: String? = null
        get() {
            if (field == null) {
                val date = dateFormat?.parse(ect ?: "")
                field = if (date != null) getSince(date, Calendar.getInstance()) else null
                Timber.i("getting since -> ${field != null}")
            }
            return field
        }
        private set//close the output stream//Then decode from the output stream and get the image.//and there is an actual LinkedFile behind the Key//If it hasn't been resolved...

    var ect: String? = null
        get() {
            if (field.isNullOrEmpty()) {
                val kmd = get(KMD_FIELD_NAME)
                field = if (kmd is Map<*, *>) {
                    val ect = kmd[Constants.ECT_FIELD_NAME]
                    ect?.toString() ?: ""
                } else {
                    null
                }
            }
            return field
        }

    /**
     * Get the thumbnail from the LinkedResource
     *
     * Note it closes the output stream.
     *
     * @return null or the image attachment
     */
    var thumbnail: Bitmap? = null
        get() {
            //If it hasn't been resolved...
            if (field == null) {
                //and there is an actual LinkedFile behind the Key
                val file = getFile(Constants.ATTACHMENT_NAME)
                if (file != null) {
                    //Then decode from the output stream and get the image.
                    val byteArr = file.output?.toByteArray()
                    field = BitmapFactory.decodeByteArray(byteArr, 0,byteArr?.size ?: 0)
                    try {
                        //close the output stream
                        file.output?.close()
                    } catch (e: Exception) {
                        Timber.e(e)
                    }
                }
            }
            return field
        }
        private set

    fun addComment(newComment: CommentEntity) {
        if (comments == null) {
            comments = ArrayList()
        }
        val ret = KinveyReference(Constants.COL_COMMENTS, newComment.id)
        comments?.add(ret)
    }

    fun resetCommentReferences() {
        if (comments == null) {
            return
        }
        for (c in comments!!) {
            c.remove("_obj")
        }
    }
}